src/odf_common_plug.c

```diff
@@ -214,25 +214,38 @@ void *odf_get_binary(char *ciphertext)
 char *odf_prepare(char *fields[10], struct fmt_main *self) {
 	if (!strncmp(fields[1], "$sxc$*", 6)) {
 		static char *buf = NULL;
-		char *cp1, *cp2, *cp3;
+		char *cp, *part1, *part2;
+		size_t len1, len2;
 		int i;
+		const size_t max_len = 3*1024;
 		if (!buf)
-			buf = mem_alloc_tiny(3*1024, 4);
-		cp1 = buf;
-		cp2 = fields[1];
-		cp2 += 6;
-		strcpy(cp1, FORMAT_TAG);
-		cp1 += strlen(FORMAT_TAG);
-		cp3 = cp2;
+			buf = mem_alloc_tiny(max_len, 4);
+		/* $sxc$*1*2*3*4*5*6*7*8*9*10*11*12 */
+		/* ^^^^^^ replace tag; remove ^^^ field #11. */
+		part1 = cp = fields[1] + 6;
 		for (i = 0; i < 10; ++i) {
-			cp3 = strchr(cp3, '*');
-			++cp3;
+			cp = strchr(cp, '*');
+			if (!cp)
+				return fields[1];
+			++cp;
 		}
-		strncpy(cp1, cp2, cp3-cp2);
-		cp1 += cp3-cp2;
-		cp3 = strchr(cp3, '*');
-		cp3++;
-		strcpy(cp1, cp3);
+		len1 = cp - part1;
+		cp = strchr(cp, '*');
+		if (!cp)
+			return fields[1];
+		++cp;
+		part2 = cp;
+		len2 = strlen(part2);
+		if (FORMAT_TAG_LEN + len1 + len2 >= max_len - 1)
+			return fields[1];
```

Comment:
Doesn't this ensure room for 1 byte more than needed? I think it should be either `>=` or `- 1`, not both.

Reply:
1 additional byte is needed for null byte.

I thought to write it as `FORMAT_TAG_LEN + len1 + len2 + 1 >= max_len`. Would it be better?

OTOH I had thoughts to write `len1 + len2 >= max_len - 1 - FORMAT_TAG_LEN`, so constants would be reduced into one. And `len1 + len2` would not overflow `size_t` even if new tag would be longer one to be replaced (currently they have same length).

Reply:
Hm, I could include `+1` into `len2`. Then null byte would be copied by `memcpy`. It would save 2 lines of code.

Somehow I thought it would be more natural with bot parts lacking null byte. Now it does not seem so natural.

Reply:
I included `+1` into `len2` and replaced commit.

Reply:
> 1 additional byte is needed for null byte.

Sure, but you left 2 bytes, not 1. I haven't reviewed your further changes here yet - I think they were unneeded.

Reply:
OK, the new changes look good to me, but you're still leaving 1 more byte unused. You can change the check from `>=` to `>`.

Reply:
Indeed. Somehow I could compute it in my head and had to draw a scheme...

Reply:
s/could/could not/